hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/ViewDistributedFileSystem.java

```diff
@@ -0,0 +1,1864 @@
+  public DFSInotifyEventInputStream getInotifyEventStream(long lastReadTxid)
+      throws IOException {
+    if (this.vfs == null) {
+      return super.getInotifyEventStream();
+    }
+    throw new UnsupportedOperationException(
+        "getInotifyEventStream is not supported in ViewDFS");
+  }
+
+  @Override
+  // DFS only API.
+  public void setErasureCodingPolicy(final Path path, final String ecPolicyName)
+      throws IOException {
+    if (this.vfs == null) {
+      super.setErasureCodingPolicy(path, ecPolicyName);
+      return;
+    }
+
+    ViewFileSystemOverloadScheme.MountPathInfo<FileSystem> mountPathInfo =
+        this.vfs.getMountPathInfo(path, getConf());
+    checkDFS(mountPathInfo.getTargetFs(), "setErasureCodingPolicy");
+    ((DistributedFileSystem) mountPathInfo.getTargetFs())
+        .setErasureCodingPolicy(mountPathInfo.getPathOnTarget(), ecPolicyName);
+  }
+
+  @Override
+  public void satisfyStoragePolicy(Path src) throws IOException {
+    if (this.vfs == null) {
+      super.satisfyStoragePolicy(src);
+      return;
+    }
+    this.vfs.satisfyStoragePolicy(src);
+  }
+
+  @Override
+  public ErasureCodingPolicy getErasureCodingPolicy(final Path path)
+      throws IOException {
+    if (this.vfs == null) {
+      return super.getErasureCodingPolicy(path);
+    }
+
+    ViewFileSystemOverloadScheme.MountPathInfo<FileSystem> mountPathInfo =
+        this.vfs.getMountPathInfo(path, getConf());
+    checkDFS(mountPathInfo.getTargetFs(), "getErasureCodingPolicy");
+    return ((DistributedFileSystem) mountPathInfo.getTargetFs())
+        .getErasureCodingPolicy(mountPathInfo.getPathOnTarget());
+  }
+
+  @Override
+  public Collection<ErasureCodingPolicyInfo> getAllErasureCodingPolicies()
+      throws IOException {
+    if (this.vfs == null) {
+      return super.getAllErasureCodingPolicies();
+    }
+    return defaultDFS.getAllErasureCodingPolicies();
+  }
+
+  @Override
+  public Map<String, String> getAllErasureCodingCodecs() throws IOException {
+    if (this.vfs == null) {
+      return super.getAllErasureCodingCodecs();
+    }
+    return defaultDFS.getAllErasureCodingCodecs();
+  }
+
+  @Override
+  public AddErasureCodingPolicyResponse[] addErasureCodingPolicies(
+      ErasureCodingPolicy[] policies) throws IOException {
+    if (this.vfs == null) {
+      return super.addErasureCodingPolicies(policies);
+    }
+    return defaultDFS.addErasureCodingPolicies(policies);
+  }
+
+  @Override
+  public void removeErasureCodingPolicy(String ecPolicyName)
+      throws IOException {
+    if (this.vfs == null) {
+      super.removeErasureCodingPolicy(ecPolicyName);
+      return;
+    }
+    defaultDFS.removeErasureCodingPolicy(ecPolicyName);
+  }
+
+  @Override
+  public void enableErasureCodingPolicy(String ecPolicyName)
+      throws IOException {
+    if (this.vfs == null) {
+      super.enableErasureCodingPolicy(ecPolicyName);
+      return;
+    }
+    defaultDFS.enableErasureCodingPolicy(ecPolicyName);
```

Comment:
I doubt this. Shouldn't enable/disable/add/remove EC Policy go to all underlying child DFS rather than defaultDFS?

Reply:
This is one point I was thinking what to do. One way I agree that running on all child hdfs-s make sense. But other I was thinking that if a user working on a specific mount and wants to call would make the call goes and disturb all other clusters also. However, Unfortunately  we don't have choice to choose specific cluster from the current API signatures.
Whats your opinion? 
Ideally they should be done from admins command line as they are not part of FileSystem API. If they are from admin, users can use -fs option to run on specific child clusters.

Reply:
Well in any case, I think making a call to `defaultDFS` isn't going to help much. Somebody does an `enablePolicy` and post that if he tries `setPolicy` which goes to another `FS` it will fail, he might try `listPolicies` which will again go to `defaultDFS` and the policy would show as `enabled` there, so this behavior would be very confusing in that case.
There are two ways only IMO, we let the admin only do it using `-fs` and throw an `UnsupportedExceptions` for these API's, 
Second solution is shooting calls to all `child FS`, This is what is done in case of `RBF` as well, I think in `ViewFileSystem` the `getAllStoragePolicies` API also does something like this. 
Personally, I feel the second option would be little better, as post that the client operations shall work without any restrictions/issues

Reply:
@ayushtkn , Thank you for your opinion. I attempted to delegate EC calls to all child filesystems. 

To the other approach the APIs delegated only defaultFS: if we don't delegate to defaultFS and simply throw USOE and if defautFS is same as the fs.defaultFS uri, then we will never be able to call this defaultFS. Because it will be resolved to same VDFS class and find mount points configured with same authority.
So, if we delegate to defaultFS in the current instance and for all other childFS, users can use -fs option. That will work as otherFS authority most probably will be different and they will be able to initialized successfully, but without any mounts.

Reply:
Thanx @umamaheswararao 
I have a small doubt-
``  // A default DFS, which should have set via linkFallback
  private DistributedFileSystem defaultDFS;``
The defaultDFS is the FS set via fallback, Correct?

`    viewFs.setSupportAutoAddingFallbackOnNoMounts(false);`
Now here we have disabled Auto adding fallback, So, in case corresponding mount entry isn't there, will a normal(read/write) call go to this defaultDFS? I think no?
* If not then why to bother `defaultDFS`, if it is not handling client calls?
* if not, Why are we not allowing fallback? is it a planned followup, or some issues with it.
* if the fallback fs is there and is present amongst as a `childFS`, We could have eliminated the `defaultDFS` logic completely? Since now a call will go to that `FS` as well?

> So, if we delegate to defaultFS in the current instance and for all other childFS

With  `and` you mean shoot calls to all childFS and then make a call to defaultDFS as well?

> defautFS is same as the fs.defaultFS uri, then we will never be able to call this defaultFS. Because it will be resolved to same VDFS class and find mount points configured with same authority.

Isn't this a client mis-config then? He configured it to get overloaded?

Well seems like, I have bothered  you too much on this, If stuff above doesn't make sense, you can have same logic as EC for `cachePool` API's as well, since `listCachePool` seems a non-Admin API. Post that you can proceed ahead with concluding this. I don't think `defaultDFS` would be bothering much in prod cases, and rest everything already is pretty cool :-)

Reply:
@ayushtkn  Thanks a lot for your time and review. You made lot of good points, no issues let's conversations going on until things are making sense. :-)
`The defaultDFS is the FS set via fallback, Correct?`
Yes, you are right.

`viewFs.setSupportAutoAddingFallbackOnNoMounts(false);
`
Why I disabled this was, in ViewDistributedFileSystem, if no mount points configured in the system, everything should work as regular DistributedFileSystem. So, we can eventually make this class (fs.hdfs.impl=ViewDistributedFileSystem) enabled by default and just don't add mount points if they don't need mounting functionality. The existing users will not see any impact as this will work same as DistributedFileSystem as the existing users would not have configured any mount points. That's the expectation here. So, if we auto add fallback, vfs#init will never fail and we always go into mount way of functionality. 
That's why every api checks vfs==null, they super.API() calls to get exactly same DistributedFileSystemFunctionality.
Hope this clears you doubt.

`Now here we have disabled Auto adding fallback, So, in case corresponding mount entry isn't there, will a normal(read/write) call go to this defaultDFS? I think no?`
Case 1: user did not configure any mounts : works same as DistributedFileSystem.
Case 2: user did configured mounts, but no fallback configured :  Whatever mount paths matching will delegate call to that fs. If no matches, it will fal with NotInMountPoint Exceptions. APIs like IsInSafeMode will fail as there is not defaultFS( that is fallback). User's make use -fs from command line and call to specific child fs. 
Case 3: user did configured mounts and as well as fallback: Now whatever paths matching will go to target fs. If no matches, then fallback. For APIs with out paths in argument like IsInSafeMode will simply make calls on that fallback fs. For the rest of other child file systems, they may need to do from command like with -fs option.

`if the fallback fs is there and is present amongst as a childFS, We could have eliminated the defaultDFS logic completely? Since now a call will go to that FS as well?`
That's why I tried make it mandatory config from user perspective. But we did not auto configure for the above reason, where we can enable by default(in future) fs.hdfs.impl pointing ViewDistributedFileSystem without any impact to existing users.


`With and you mean shoot calls to all childFS and then make a call to defaultDFS as well?`
What I meant was: example: you have 
fs.defaultFS = hdfs://ns1
 fallback(defaultFS) -> hdfs://ns1
/user --> hdfs://ns2

when you can isInSafeMode, you will get result from default cluster(fallback).
To get from other child's, you will get FileSystem.get(otherClusterURI, conf).isInSafeMode().

Incase if we simply throw USOE without giving from defaultCluster, we can never call to default cluster because FileSystem.get(fallbackUri, conf).isInSafeMode() will always gets USOE because the fallback and fs.defaultFS uris are same. 

`Isn't this a client mis-config then? He configured it to get overloaded?`
One of the important use case is, if you have existing cluster and you want to add mount points with respective to that cluster. Then you will that existing cluster as fallback and you will continue to use same fs.defaultFS.
So, all ops goes to your default cluster except the paths matching to mount points. That way existing users need not change uri, but they can simply add mounts with respective to that cluster by simply adding current cluster as fallback.


`you can have same logic as EC for cachePool API's as well, since listCachePool seems a non-Admin API.`
I am thinking in similar lines. Let me add cachePool API shoot to all fss.

BTW, please note currently getChildFileSystems does not include, fallback fs. It will be fixed with HDFS-15529    
Thanks again for your review.

Reply:
Thanx for the details. Makes sense.

Reply:
Thanks